tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12930,6 +12930,40 @@ bool ValidateAttributeTargetIsFunction(Sema& S, Decl* D, const AttributeList &A)
   return false;
 }
 
+void Sema::ValidateShaderAttributes(Decl *D, const AttributeList *A) {
+  std::string lastAttrStr = "";
+  const AttributeList *lastAttr = nullptr;
+
+  for (const AttributeList *l = A; l; l = l->getNext()) {
```

Comment:
A key flaw of this implementation is this loop. You're looping over the attribute list _inside_ a loop over the attribute list. So this is `n^2`, _and_ you risk duplicating checks and diagnostics.

The intent of `HandleDeclAttributeForHLSL` is to handle a single attribute, not a list of them.